### PR TITLE
Create transaction

### LINKED ISSUE #547: When a transaction is made, that category's market share should increase

### LINKED ISSUE #548: Create Transaction needs to be rewritten and unit tested

### PR DESCRIPTION
**Pull request by ritmatter**
Fixes #548
Fixes #547

The new transaction post route is not actually unit tested very well, since I had a really hard time mocking the chains of .then(). But, I think its satisfactory for now, since the infrastructure is now much clearer and more strongly unit tested

### CODE REVIEW

File: api/handlers/transaction.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        return res.status(400).send(err);
+      }).then(function(fromuser) {
+        fromUser = fromuser;
+        return categoryPromise;
+      }, function(err) {
+        winston.log('error', 'Error creating transaction: %s', err.toString());
+        return res.status(400).send(err);
+      }).then(function(categoryP) {
+        category = categoryP;
+        // Update fields for all of the documents a they should be for the transaction
```
Comment (krazemon) [inline api/handlers/transaction.js]: Typo: s/documents a/documents as/

File: api/routes/utils.js [hunk 2]
```diff
@@ -506,13 +548,13 @@ var utils = {
...
-      return null;
+      return 'User is not an investor for this category';
     }
 
     // Add the investment to the portfolio
     if (amount > 0) {
-      var percentage = Number(amount/toUserCategoryTotal);
-      var dividend   = Math.round(percentage * toUserCategoryTotal * DIVIDEND_RATE * 100) / 100;
+      var percentage = Number(amount/toUserReps);
+      var dividend   = Math.round(percentage * toUserReps * DIVIDEND_RATE * 100) / 100;
```
Comment (krazemon) [inline api/routes/utils.js]: So was this wrong before?

File: api/routes/utils.js [hunk 3]
```diff
@@ -710,26 +752,50 @@ var utils = {
...
 
-  // Given a category name, update the percentiles and percentages, for all investors
-  updateInvestors: function(category, cb, username, expertCategoryTotal) {
+  // Given a category name, update the percentiles for all the experts and investors in that category
+  // In addition, update the investor dividends
+  updatePercentilesAndDividends: function(categoryName, expert, cb) {
     var self = this;
-    var investorsPromise = User.findInvestorByCategory(category, function() {});
+    var i = self.getCategoryIndex(expert, categoryName);
+    if (i === -1) {
```
Comment (krazemon) [inline api/routes/utils.js]: Can this just be ```if i < 0```? If you don't want to change it it's fine. Just seems like a safer convention to use in general.

File: spec/routes_utils_spec.js [hunk 4]
```diff
@@ -411,93 +414,139 @@ describe('Utils: ', function() {
     });
   });
 
-  describe('updateInvestors: ', function() {
+  describe('updatePercentilesAndDividends: ', function() {
+    var cb, realUpdateExpertPercentiles, realUpdateInvestorPercentiles, expert, categoryName;
+
     beforeEach(function() {
       cb = jasmine.createSpy();
```
Comment (krazemon) [inline spec/routes_utils_spec.js]: Neat trick.

File: spec/transaction_handler_spec.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      end: jasmine.createSpy()
+    };
+  });
+
+  afterEach(function() {
+    expect(res.status.callCount).toEqual(1);
+    expect(res.send.callCount).toEqual(1);
+  });
+
+  describe('transactions: ', function() {
```
Comment (krazemon) [inline spec/transaction_handler_spec.js]: So this only tests the error case?

**Comment by krazemon** [comment]
LGTM. There's one typo but it's your call whether you want to fix it. Also, if you could answer my questions that would be great just for my own knowledge.

File: api/routes/utils.js (hunk 2 above)
Comment (ritmatter) [inline api/routes/utils.js]: Just a refactoring, it wasn't wrong

File: api/routes/utils.js (hunk 3 above)
Comment (ritmatter) [inline api/routes/utils.js]: It can be. I'll change it here.

File: spec/transaction_handler_spec.js (hunk 5 above)
Comment (ritmatter) [inline spec/transaction_handler_spec.js]: This only tests invalid inputs. I couldn't figure out how to mock the promise chain.

File: api/handlers/transaction.js (hunk 1 above)
Comment (ritmatter) [inline api/handlers/transaction.js]: fixed

**Pull request merged**